### ISSUE TITLE
Automatic merge for the snapshot branch

### ISSUE THREAD

**Issue by mtojek** [elastic/package-storage#592]
As we agreed today, we have to enabled automatic merges in the package-storage repository for the snapshot branch. 

Enabling it for staging is a bit more complicated as it requires more logic (promotion concern two stages and the Docker build order is important).

Let's discuss ideas.

1. New scheduled (every 15 minutes?) Jenkins job responsible for merging changes that looks good.
2. Artificial CI check to fire the merge action. 
3. Github action to automerge if the CI job is green.

/cc @ruflin @ycombinator 

/cc @kuisathaverat (feel free to share your experience as you're modifying the release pipeline at the moment)

**Comment by ruflin** [comment]
I think we need to differentiate two different PR's here. The ones that just change a package and should be auto merged. They are also generated by a script. And then the manual PR's like updating a registry which we might want to have a discussion on. I guess we can use labels to make this differentiation?

Not sure I fully understand how having automatic merge by CI in staging is different from snapshot or manual if it is green?

**Comment by mtojek** [comment]
> I think we need to differentiate two different PR's here. The ones that just change a package and should be auto merged. They are also generated by a script.

Yes. The CI Jenkins builder runs `mage UpdatePackageStorage` which opens PRs against the snapshot branch (user Elastic Machine). The question is what is the reasonable design of the auto merge?

> And then the manual PR's like updating a registry which we might want to have a discussion on. I guess we can use labels to make this differentiation?

I assume you're referring to manual version updates of the package-registry (bumping up registry release). I would leave this as manual actions.

> Not sure I fully understand how having automatic merge by CI in staging is different from snapshot or manual if it is green?

It is different because of the dependency between stages. PRs to the snapshot branch are created using a script (adding new files), but PRs to the staging and the production are created with the `elastic-package promote` (moving packages around). If you promote package between staging and production, there are two PRs created - in this case merging order is important as we have to wait not only for the merge to the destination branch, but also for the relevant Docker image to be pushed. 
Consider also the fact that in the mean time, there might be two independent merge-jobs running in background, which might lead to a race condition.

**Comment by ruflin** [comment]
Got it, missed the two PR parts. But I'm wondering if it matters if we look at staging and snapshot independent. User promotes package to staging. Now a script in staging decides if the added package can be merged and a script in snapshot decides if the removal is ok. We now get into the situation, that it was removed from snapshot but not added to staging was red. What if we just say this is acceptable?

**Comment by mtojek** [comment]
> We now get into the situation, that it was removed from snapshot but not added to staging was red. What if we just say this is acceptable?

What if it's the system package, which is crucial and it silently disappears. Keep in mind that there might be a different job which can also approve different PR, and another developer will rollout deployment with impaired snapshot image.
I'm not quite sure if we can accept this risk, I'm afraid we might have missed some edge cases. It's safer to wait for the staging PR to get merged.

**Comment by kuisathaverat** [comment]
we can have a stage at the end of the pipeline that merges the PR if the build is successful and is labeled with auto-merge (or something else), this merge and squash the PR, and a new build is triggered because of the merge on the snapshot branch.

Reading your comments, I wonder if there is a case where we want to auto promote a merge on the snapshot branch to the stage environment.

**Comment by mtojek** [comment]
> Reading your comments, I wonder if there is a case where we want to auto promote a merge on the snapshot branch to the stage environment.

It depends on the package owners whether they prefer to keep their changes in snapshot or move quickly to staging. Based on observations, the staging is used mostly during the testing phase, as these packages are candidates for prod release.

Considering the fact that promotion involves two PRs (add and remove), what does the auto-promote look like? To be clear, I think we don't need it at the moment. With the enabled auto-promotion to staging branch, we can say that we don't need snapshot at all :)

**Comment by mtojek** [comment]
As discussed offline with @kuisathaverat , it will be covered by the robots team.

**Comment by mtojek** [comment]
@kseniia-kolpakova @kuisathaverat Hi! Do you have any ETA for this issue? I would be nice to see this feature live.

**Comment by kuisathaverat** [comment]
we are checking a couple of options mergify and the new automerge feature of GitHub, we want to test it on our repo for some time and then decided.

**Comment by kuisathaverat** [comment]
I want to start to work on this, I understand that is only for the branch snapshot at the moment, so `Github action to automerge` cannot be used because it is enabled at the repository level so affects all branches. If the logic to merge is simple we can go with something in the pipeline directly. I'll make the following proposals:

* Automerge when the CI pass
* Automerge when the CI pass and the PR is approved
* Automerge when the CI pass and the PR has a label auto-merge
* Automerge when the CI pass, the PR has a label auto-merge, and the PR is approved

if the logic is more complex we will go for mergify

**Comment by ruflin** [comment]
I like the idea of the label + auto-merge. This prevents any random PR's to be merged and also AFAIK the community cannot add labels so nothing gets merged accidentally. @mtojek I assume adding the label `auto-merge` to the PR's by the script shouldn't be an issue? My vote is for option 3.

**Comment by mtojek** [comment]
It would great to use the same logic for staging and prod too. I assume that when somebody promotes their changes between stages, they expect to merge them eventually.

I can give your more rules if you plan to cover also other stages:
* one concurrent merge operation the entire package-storage - keep in mind that promotion consists of a pair of actions, "promote" and "delete" PRs
* would be good to wait fo the CI job to finish building new image
* promotion: it's be forbidden to merge "delete" PR  if "promote" PR is not merged, it means that the package will disappear :)
* built images depends on each other, so consider build job triggers

>  @mtojek I assume adding the label auto-merge to the PR's by the script shouldn't be an issue?

No, the script is here: https://github.com/elastic/integrations/tree/master/dev/update-package-storage

**Comment by kuisathaverat** [comment]
If we use label we can use the auto merge from GitHub, we create a stage in the Jenkins Pipeline that creates a check if the label is there, then we add that check to the auto merge.

>would be good to wait fo the CI job to finish building new image

we can restrict the number of concurrent jobs to one

>one concurrent merge operation the entire package-storage - keep in mind that promotion consists of a pair of actions, "promote" and "delete" PRs
>promotion: it's be forbidden to merge "delete" PR if "promote" PR is not merged, it means that the package will disappear :)

I am not familiar with that process, the "promote" PR looks like we can auto merge it. The "delete" PR can be marked as auto merge after the "promote" PR is merged. The key here is to identify those merge commits.

this would be more or less the process
* we have the "promote" PR with the title "promote: new version of apache package 1.3" is marked as auto merge
* we have the "delete" PR with the title "delete: remove apache package 1.3" is NOT marked as auto merge
* the "promote" PR pass the CI and match the condition we set to auto merge (CLA+CI+2 approvals for example)
* then the "promote" PR is merged
* A new build in the branch is triggered 
* The CI pass and detects that the commit is a "promote:" commit so the CI mark the "delete" PR as auto merged
* the next stage branch build is triggered (IIRC snapshot -> staging and production)
* The "delete" PR is auto merged
* A new build in the branch is triggered 
* the next stage branch build is triggered (IIRC snapshot -> staging and production)

>built images depends on each other, so consider build job triggers

if the merge is in cascade the build is in the same order

**Comment by ruflin** [comment]
To keep the initial implementation simpler, perhaps we can do the auto merge only for the new and not the delete?

**Comment by mtojek** [comment]
> perhaps we can do the auto merge only for the new and not the delete

I don't agree with idea. "Promote" and "Delete" must go in a transaction, otherwise we'll end up with a mess in the storage. Some people will remember about the cleanup, other won't.

If this is about defining go/no-go rules for merging operation, for "Delete" action we should check if the package is available in the next stage using public the EPR endpoint.

**Comment by kuisathaverat** [comment]
take a look at our tests with mergify https://github.com/elastic/apm-pipeline-library/pull/952/checks

**Comment by cachedout** [comment]
@mtojek Did you have a chance to review the Mergify checks that @kuisathaverat [suggested earlier](https://github.com/elastic/package-storage/issues/592#issuecomment-774108479)?

As he points out, we're using Mergify on a few other projects right now and have been quite satisfied. If you're all right with it, I'll put together a proposed configuration for your review.

**Comment by cachedout** [comment]
There's been a fair amount of discussion on this thread covering a variety of use cases, but it seems as if we might be able to start with the a small amount of automation to cover the most common workflow and then expand from there.

Just as a starting point for discussion, what about something like this?

```
pull_request_rules:
  - name: Automatic merge for snapshot branch when CI passes and author is elasticmachine
    conditions:
      - check-success=package-storage/pr-merge
      - base=snapshot
      - author=elasticmachine
    actions:
      merge:
        method: merge
  - name: Automatic merge for snapshot branch when CI passes and the author is inside Elastic and there is one approval
    conditions:
      - check-success=package-storage/pr-merge
      - base=snapshot
      - author="@elastic/observability"
      - "#approved-reviews-by>=2"
    actions:
      merge:
        method: merge
```

As you can see, there are two rulesets both of which are hopefully self-explanatory. Though there was [an earlier discussion of a label ](https://github.com/elastic/package-storage/issues/592#issuecomment-771612321), I didn't see the need for it since we can just check to see if the PR comes from the `elasticmachine` user and in that way be assured that it's the result of a CI job. (Though, perhaps that is not strict enough? I might be missing something here.)

Then there is a second case which is essentially, anybody on the Elastic Observability team who submits a PR which passes tests and then also has two reviews. 

For reference documentation, [please see here](https://docs.mergify.io/), but I'm hoping this is enough of a starting point that we can refine from here. Please let me know what you think.

**Comment by mtojek** [comment]
@cachedout 

I'm not sure about the second case, did you have any particular pull request in mind?

Regarding the first rule,  please go for it, it would give us a real value as most of devs forget about the first step which is "parking" the integration in the package-storage. Later on we can iterate on the promote/remove PRs.

BTW I like the idea of a special label ("auto-merge", "mergify"). In case of a disaster we can quickly disable the mechanism.

**Comment by cachedout** [comment]
@mtojek Thanks for the reply. A few more brief questions.

1. Some of these PRs submitted by Elastic Machine are marked for review by certain teams. [First example](https://github.com/elastic/package-storage/pull/933) and [second example](https://github.com/elastic/package-storage/pull/938). Should they be automatically merged before reviews are completed?

2. Regarding labels, it looks like the window between a PR being submitted and tests completing [is ~16min](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fpackage-storage/detail/PR-938/2/pipeline). If we have the CI affix a label which could then be removed in the event of an emergency, is that enough time for you to detect the issue and respond?

**Comment by mtojek** [comment]
> Some of these PRs submitted by Elastic Machine are marked for review by certain teams. First example and second example. Should they be automatically merged before reviews are completed?

Yes, if the CI is happy, we can merge them. Teams are marked as reviewers for visibility.

> Regarding labels, it looks like the window between a PR being submitted and tests completing is ~16min. If we have the CI affix a label which could then be removed in the event of an emergency, is that enough time for you to detect the issue and respond?

SGTM!

**Comment by cachedout** [comment]
I've tried to enable the Mergify integration [using these instructions](https://docs.mergify.io/getting-started/) but every time I try to add the this repo, the change seems to have no effect.

@elastic/observablt-robots when adding the other repos, did you have to do anything special?

**Comment by v1v** [comment]
> @elastic/observablt-robots when adding the other repos, did you have to do anything special?

There are two steps:
- Create the `mergify.yml` file
- Raise an issue in the infra repo and poke the infra/hosting team to ask them to enable the mergify app for this repo.

Once it's enabled and the mergify PR is merged then things should be working as expected.

**Comment by cachedout** [comment]
We have enabled Mergify on the repo and I worked with @mtojek today to successfully test and verify the implementation. We[ submitted a PR ](https://github.com/elastic/package-storage/pull/1162)and Mergify ran the requisite checks and merged it as intended.

As a side-note, enabling the Mergify integration on this repo also merged some _existing_ PRs, which was somewhat unexpected. We have [started a discussion with the Mergify developers regarding this](https://github.com/Mergifyio/mergify-engine/discussions/2412).

Per the Slack discussion with @mtojek , I am going to go ahead and resolve this issue. Thanks!
